File: ios/JACL/GameMap.swift
//  GameMap.swift
//  The explored map parsed from mapping.library's <jacl-map> data block.

import Foundation

/// A room node: (x,y) is the top-left in the web map's pixel space
/// (120 px/cell, 80 px nodes); `here` marks the current room.
struct MapNode { let x, y: Int; let here: Bool; let label: String }

/// An exit line between two room centres; `updown` draws an up/down arrow.
struct MapEdge { let x1, y1, x2, y2: Int; let updown: Bool }

/// The explored map: canvas `w`x`h` (px) plus nodes and edges.
struct GameMap { let w, h: Int; let nodes: [MapNode]; let edges: [MapEdge] }

/// Parse the lines between `<jacl-map>` and `</jacl-map>` (records "M w h",
/// "E x1 y1 x2 y2 ud", "N x y here label…"). nil if nothing is drawable.
func parseGameMap(_ lines: [String]) -> GameMap? {
    var w = 0, h = 0
    var nodes: [MapNode] = []
    var edges: [MapEdge] = []
    for raw in lines {
        let s = raw.trimmingCharacters(in: .whitespaces)
        if s.hasPrefix("M ") {
            let p = s.split(separator: " ")
            if p.count >= 3 { w = Int(p[1]) ?? 0; h = Int(p[2]) ?? 0 }
        } else if s.hasPrefix("E ") {
            let p = s.split(separator: " ")
            if p.count >= 6, let x1 = Int(p[1]), let y1 = Int(p[2]),
               let x2 = Int(p[3]), let y2 = Int(p[4]) {
                edges.append(MapEdge(x1: x1, y1: y1, x2: x2, y2: y2, updown: p[5] == "1"))
            }
        } else if s.hasPrefix("N ") {
            // "N x y here label…" -- label (rest) may contain spaces.
            let p = s.split(separator: " ", maxSplits: 4, omittingEmptySubsequences: true)
            if p.count >= 5, let x = Int(p[1]), let y = Int(p[2]) {
                nodes.append(MapNode(x: x, y: y, here: p[3] == "1", label: String(p[4])))
            }
        }
    }
    guard !nodes.isEmpty else { return nil }
    return GameMap(w: max(w, 1), h: max(h, 1), nodes: nodes, edges: edges)
}

File: android/app/src/main/java/au/com/dangarmarine/jacl/GameMap.kt
package au.com.dangarmarine.jacl

/** A room on the map. (x,y) is the node's top-left in the web map's pixel
 *  space (120 px/cell, 80 px nodes); [here] marks the current room. */
data class MapNode(val x: Int, val y: Int, val here: Boolean, val label: String)

/** An exit line between two room centres; [updown] draws an up/down arrow. */
data class MapEdge(val x1: Int, val y1: Int, val x2: Int, val y2: Int, val updown: Boolean)

/** The explored map: canvas size [w]x[h] (px) plus nodes and edges. */
data class GameMap(val w: Int, val h: Int, val nodes: List<MapNode>, val edges: List<MapEdge>)

/**
 * Parse a `<jacl-map>` data block (the lines between the markers) emitted by
 * mapping.library's +map_native. Records: "M w h", "E x1 y1 x2 y2 ud",
 * "N x y here label...". Returns null if there's nothing drawable.
 */
fun parseGameMap(lines: List<String>): GameMap? {
    var w = 0; var h = 0
    val nodes = ArrayList<MapNode>()
    val edges = ArrayList<MapEdge>()
    for (line in lines) {
        val s = line.trim()
        when {
            s.startsWith("M ") -> {
                val p = s.split(' ')
                w = p.getOrNull(1)?.toIntOrNull() ?: 0
                h = p.getOrNull(2)?.toIntOrNull() ?: 0
            }
            s.startsWith("E ") -> {
                val p = s.split(' ')
                if (p.size >= 6) edges.add(MapEdge(
                    p[1].toInt(), p[2].toInt(), p[3].toInt(), p[4].toInt(), p[5] == "1"))
            }
            s.startsWith("N ") -> {
                // "N x y here label..." -- label is the rest (may contain spaces).
                val p = s.split(' ', limit = 5)
                if (p.size >= 5) nodes.add(MapNode(
                    p[1].toInt(), p[2].toInt(), p[3] == "1", p[4]))
            }
        }
    }
    if (nodes.isEmpty()) return null
    return GameMap(if (w > 0) w else 1, if (h > 0) h else 1, nodes, edges)
}
